base/c.jl

```diff
@@ -503,3 +503,156 @@ end
+    end
+
+    # get the function symbols
+    func = let f = call.args[1]
+        if Meta.isexpr(f, :.)
+            :(($(f.args[2]), $(f.args[1])))
+        elseif Meta.isexpr(f, :$)
+            f
+        elseif f isa Symbol
+            QuoteNode(f)
+        else
+            throw(ArgumentError("@ccall function name must be a symbol, a `.` node (e.g. `libc.printf`) or an interpolated function pointer (with `\$`)"))
+        end
+    end
+
+    # detect varargs
+    varargs = nothing
+    argstart = 2
+    callargs = call.args
+    if length(callargs) >= 2 && Meta.isexpr(callargs[2], :parameters)
+        argstart = 3
+        varargs = callargs[2].args
+    end
+
+    # collect args and types
+    args = []
+    types = []
+
+    function pusharg!(arg)
+        if !Meta.isexpr(arg, :(::))
+            throw(ArgumentError("args in @ccall need type annotations. '$(repr(arg))' doesn't have one."))
+        end
+        push!(args, arg.args[1])
+        push!(types, arg.args[2])
+    end
+
+    for i in argstart:length(callargs)
+        pusharg!(callargs[i])
+    end
+    # add any varargs if necessary
+    nreq = 0
+    if !isnothing(varargs)
+        if length(args) == 0
+            throw(ArgumentError("C ABI prohibits vararg without one required argument"))
+        end
+        nreq = length(args)
+        for a in varargs
+            pusharg!(a)
+        end
+    end
+
+    return func, rettype, types, args, nreq
+end
+
+
+function ccall_macro_lower(convention, func, rettype, types, args, nreq)
+    lowering = []
+    realargs = []
+    gcroots = []
+
+    # if interpolation was used, ensure  variable is a function pointer at runtime.
+    if Meta.isexpr(func, :$)
+        push!(lowering, Expr(:(=), :func, esc(func.args[1])))
+        func = :func
+        check = quote
+            if !isa(func, Ptr{Nothing})
+                name = $(QuoteNode(func))
+                throw(ArgumentError("interpolated function `$name` was not a Ptr{Cvoid}, but $(typeof(func))"))
+            end
+        end
+        push!(lowering, check)
+    else
+        func = esc(func)
+    end
+
+    for (i, (arg, type)) in enumerate(zip(args, types))
+        sym = Symbol(string("arg", i, "root"))
+        sym2 = Symbol(string("arg", i, ))
+        earg, etype = esc(arg), esc(type)
+        push!(lowering, :($sym = Base.cconvert($etype, $earg)))
+        push!(lowering, :($sym2 = Base.unsafe_convert($etype, $sym)))
+        push!(realargs, sym2)
+        push!(gcroots, sym)
+    end
+    etypes = Expr(:call, Expr(:core, :svec), types...)
+    exp = Expr(:foreigncall,
+               func,
+               esc(rettype),
+               esc(etypes),
+               nreq,
+               QuoteNode(convention),
+               realargs..., gcroots...)
+    push!(lowering, exp)
+
+    return Expr(:block, lowering...)
+end
+
+"""
+    @ccall some_c_function(arg::Type [...])::ReturnType
+
+convert a julia-style function definition to a ccall:
```

Comment:
I suspect people will generally want to use `@ccall` rather than `ccall`, so I suggest we keep these docs mostly self-contained (perhaps with a "See also: `ccall`"). Copying and adapting the `ccall` docs for the style and formal description would make sense I think, followed by an `# Examples` section with the examples you've got here. (For me, the examples here already make this more helpful than the current `ccall` docs.)

Reply:
I'll try to wrap this up this weekend. Speaking of documentation, I guess I need to add a news item somewhere? I think I remember reading that.

Also, I guess there also need to be a discussion about including `@ccall` in the regular documentation, but I guess others will have opinions about this topic.

Reply:
Yes, news please!

I do think it will be natural to completely replace `ccall` with `@ccall` in julia 2.0 (no point having two ways to do the same thing). So with that in mind the manual section should be updated. But I also think that could be a separate PR.

Reply:
Resolved in the following flurry of janky commits.